Part 1: Block attribute(s) authoring in weaker layer(s)

HamedSabri-adsk opened the pull request:
**Part 1:** Initial PR to add support for blocking "Transformation attribute authoring via manipulators" and "Single attribute authoring" in weaker layers.

e.g Single Attribute:

![single_attr_block](https://user-images.githubusercontent.com/48299423/111213087-ca4c5e80-85a6-11eb-9deb-fae5e52d84a3.gif)

e.g Transformation via manipulators:

![manipulation_attr_block](https://user-images.githubusercontent.com/48299423/111213100-cf111280-85a6-11eb-863e-d2dbcda6bdf8.gif)

HamedSabri-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -191,6 +216,53 @@ void applyCommandRestriction(const UsdPrim& prim, const std::string& commandName
+    // composition arc(s). Some more details can be found:
+    // https://groups.google.com/g/usd-interest/c/xTxFYQA_bRs/m/lX_WqNLoBAAJ
+    SdfLayerHandle strongLayer;
+    for (const auto& spec : propertyStack) {
```
@ppt-adsk See the comments in code that I left here.

HamedSabri-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -191,6 +216,53 @@ void applyCommandRestriction(const UsdPrim& prim, const std::string& commandName
+            attr.GetBaseName().GetText(),
+            strongLayer->GetDisplayName().c_str());
+
+        MGlobal::displayError(err.c_str());
```
I originally intended to just "throw std::runtime_error" or "throw" but that won't work considering the very long
error message that we would get when the exceptions are thrown.

`# Error: RuntimeError: file <string> line 2: RuntimeError: file C:\Users\sabrih\maya\builds\master\maya\build\RelWithDebInfo\runTime\Python\lib\site-packages\maya\internal\ufeSupport\ufeCmdWrapper.py line 32: Cannot edit [height] attribute because there is a stronger opinion in [anonymousLayer4].`

Instead this function returns a bool and simply shows an error message using MGlobal::displayError

HamedSabri-adsk replied:
@ppt-adsk Good point. addressed in fe8131c1c837ebf7bf5405e7c594df7b5b7b1db4

HamedSabri-adsk commented on lib/mayaUsd/ufe/wrapUtils.cpp:
```diff
@@ -194,6 +217,47 @@ int ufePathToInstanceIndex(const std::string& ufePathString)
+        return false;
+    }
+
+    return true;
```
I duplicated the isAttributeEditAllowed() routine here to make it work with the python binding. 
I realize it is a bit wacky but having a python binding for "Ufe/Private/Utils" turned out to be more work. We can chat offline if you think there is a better way to go about this.

ppt-adsk replied:
Yeah, the code duplication is unfortunate.  Would it help if you moved it out of private/Utils and put it somewhere else?  Because now, your Python code isn't testing your C++ code...

HamedSabri-adsk replied:
@ppt-adsk I got rid of the wacky duplicate by moving out the code one level up into "Ufe/Utils" which would work better.

Addressed in : fe8131c1c837ebf7bf5405e7c594df7b5b7b1db4

ppt-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -21,13 +21,37 @@
+{
+    const auto& prim = attr.GetPrim();
+    const auto& stage = prim.GetStage();
+    const auto& layerStack = stage->GetLayerStack();
```
This is all fine, of course --- but I wonder performance-wise.  Unless the layer stack changes (and I think there are USD events to detect this --- maybe this one
https://graphics.pixar.com/usd/docs/api/class_sdf_notice_1_1_layers_did_change.html
but the docs appear broken), you can compute the index once and cache it.  Maybe we can wait for this optimization until we figure out what Pixar will be giving us (or when they will give it to us :) ).

I can see something like
std::unordered_map<SdfLayerHandle, int> layerIndices;

HamedSabri-adsk replied:
@ppt-adsk caching the index compute should be trivial. We can add it at any-times if performance becomes a bottleneck. 

Honestly, how many layers are we talking about? Even with 500s layers which is a lot of layers to me we should have a very fast look up. 

My hope with Pixar is that the client(s) of such APIs shouldn't even think about such details. It should all happen behind the scene.

ppt-adsk replied:
I'm less concerned about the number of layers than about the number of attributes to check...  Which might be hundreds of thousands.  But you're right, let's put that on the side for now, we can revisit this later.

ppt-adsk commented on lib/mayaUsd/ufe/UsdAttribute.cpp:
```diff
@@ -81,7 +82,8 @@ template <typename T> bool setUsdAttr(const PXR_NS::UsdAttribute& attr, const T&
     // single UFE attribute changed notification is generated.
     MayaUsd::ufe::AttributeChangedNotificationGuard guard;
-    return attr.Set<T>(value);
+    bool isSetAttrAllowed = MayaUsd::ufe::isAttributeEditAllowed(attr);
```
As per my proposal, here I'd pass in a pointer to string to receive an eventual error message, and displayError() if the return value is false.

HamedSabri-adsk replied:
Addressed in fe8131c1c837ebf7bf5405e7c594df7b5b7b1db4

ppt-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -21,13 +21,37 @@
+            return layer == l;
+        });
+
+    if (iter != layerStack.end()) {
```
Bit awkward that if iter == layerStack.end() we'll return position == 0.  Should we use TF_AXIOM(iter != layerStack.end())?

ppt-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -217,13 +220,17 @@ def testAttributeBool(self):
-        ufeAttr.set(not ufeAttr.get())
+        # check to see if the attribute edit is allowed
+        isAttrEditAllowed = mayaUsdUfe.isAttributeEditAllowed(usdAttr)
+        self.assertFalse(isAttrEditAllowed, False)
```
This is weird...  assertFalse() takes a single argument...  And if you're expecting no edits allowed, then line 226 is a bit pointless :)  Same in a bunch of places.

HamedSabri-adsk replied:
@ppt-adsk  Good catch. Copy-paste damage  :) 

Addressed in 7be71c51668f70493e0533da23f91b1368e73bad

I agree that the check after is useless by my thinking was the code was still valuable to have around. I can remove them if it makes things more clear.

ppt-adsk replied:
As mentioned in the code, I'm concerned that we've lost code coverage because of this.  Would be nice to rework the test file perhaps, so that we can keep the tests?

ppt-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -613,9 +644,139 @@ def testZAttrChangeRedoAfterPrimCreateRedo(self):
+        # create a prim. This creates the primSpec in the root layer.
+        proxyShapeContextOps.doOp(['Add New Prim', 'Capsule'])
+
+        # create a USD prim
```
More like "Get the USD prim from the scene item."

ppt-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -613,9 +644,139 @@ def testZAttrChangeRedoAfterPrimCreateRedo(self):
+        radiusAttrUsd = capsulePrim.GetAttribute('radius')
+
+        # authoring new attribute edit is expected to be allowed.
+        self.assertTrue(mayaUsdUfe.isAttributeEditAllowed(radiusAttrUsd), True)
```
assertTrue() takes a single argument.

ppt-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -613,9 +644,139 @@ def testZAttrChangeRedoAfterPrimCreateRedo(self):
+        cmds.mayaUsdEditTarget(proxyShapePath, edit=True, editTarget=subLayerC)
+
+        # authoring new transformation edit is not allowed.
+        self.assertFalse(mayaUsdUfe.isAttributeEditAllowed(translateAttr), False)
```
Might want to write a value different from the existing one and confirm that the write did not happen.

HamedSabri-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -191,6 +216,53 @@ void applyCommandRestriction(const UsdPrim& prim, const std::string& commandName
+
+        // skip if the edit target layer is stronger than the propSpec layer
+        const auto propSpecLayerIndex = findLayerIndex(attr, spec->GetLayer());
+        if (targetLayerIndex <= propSpecLayerIndex) {
```
@ppt-adsk I kind of see what you mean by eliminating the need to iterate the PropertyStack() but I think the algorithm is more obvious by using it. We can discuss it if my comment is not clear. Correct, a strong layer have a lower index than a weak layer. 

The PropertyStack() would eventually go away if we decide to rely on the "site" and iterating PcpPrimIndex.  

PcpPrimIndex is an index of the all sites of scene description that contribute opinions to a specific prim, under composition semantics.

https://graphics.pixar.com/usd/docs/api/class_pcp_prim_index.html#details

HamedSabri-adsk commented on lib/mayaUsd/ufe/private/Utils.cpp:
```diff
@@ -21,13 +21,37 @@
+
+    uint32_t position { 0 };
+
+    auto iter = std::find_if(
```
I believe std::find_if is what I wanted to use there since "std::find_if" takes a unary predicate and checks whether the predicate returns true or false to a given element. However you are right, the conditional find could be replace by this:

auto iter = std::find(std::begin(layerStack), std::end(layerStack), layer);

Both ways would work. Addressed in fe8131c1c837ebf7bf5405e7c594df7b5b7b1db4

HamedSabri-adsk commented on lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp:
```diff
@@ -101,7 +102,7 @@ namespace ufe {
 
 namespace {
```
@ppt-adsk  Unrelated to this PR but I believe having an anonymous namespace inside MAYAUSD_NS_DEF::ufe namespace is very confusing. Considering there is already an anonymous namespace starting at line 40 and ending at line 98.

ppt-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -80,6 +80,24 @@ bool stringBeginsWithDigit(const std::string& inputString)
 }
 
+// find position index for a given layer in the local LayerStack ( strong-to-weak order )
+uint32_t findLayerIndex(const PXR_NS::UsdAttribute& attr, const PXR_NS::SdfLayerHandle& layer)
```
Just noticed this (sorry...), and not a big deal because it's in the unnamed namespace, but interface is awkward: you require an attr, but in fact you really want just a stage, because you use the attr only to get the stage.  If we ever factor this out, I would change this to accept a stage as the first argument.

HamedSabri-adsk replied:
addressed ac2bb4dc732e9862957b25a17eb9f6f7719adcf3

ppt-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -384,6 +402,54 @@ TfTokenVector getProxyShapePurposes(const Ufe::Path& path)
+    // opinions for this property.
+    const auto& propertyStack = attr.GetPropertyStack();
+
+    // HS March 15th,2021
```
As I mentioned I think the loop is an over-elaboration.  What you want is to compare the layer at the top of the property stack to the target layer index, something like this:

SdfLayerHandle topLayer = propertyStack.front().GetLayer();
if (findLayerIndex(attr, topLayer) < targetLayerIndex) {
    std::string err = TfStringPrintf( // etc.

HamedSabri-adsk replied:
I replaced the loop if it makes logics easier to follow.

addressed ac2bb4dc732e9862957b25a17eb9f6f7719adcf3

ppt-adsk replied:
Thanks, makes more sense to me at least :)

HamedSabri-adsk commented on lib/mayaUsd/ufe/wrapUtils.cpp:
```diff
@@ -219,43 +196,8 @@ int ufePathToInstanceIndex(const std::string& ufePathString)
-
-    return true;
+    std::string dummyStr;
+    return ufe::isAttributeEditAllowed(attr, &dummyStr);
```
That's what I tried first but boost python binding doesn't like it. 

```
 Stack trace:
15:   mayaUsd.dll!std::basic_string<char,std::char_traits<char>,std::allocator<char> >::assign Line 2659
15:   mayaUsd.dll!MayaUsd_v0::ufe::isAttributeEditAllowed Line 435
15:   _ufe.pyd!boost::python::objects::caller_py_function_impl<boost::python::detail::caller<bool (__cdecl*)(pxrInternal_v0_21__pxrReserved__::UsdAttribute const &),boost::python::default_call_policies,boost:
```

ppt-adsk replied:
That's very strange, I think the fact you have Boost in the call stack is simply because it's being called from Python.  This should work, let me know if you'd like us to debug this together.

HamedSabri-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -222,7 +222,7 @@ def testAttributeBool(self):
         # check to see if the attribute edit is allowed
         isAttrEditAllowed = mayaUsdUfe.isAttributeEditAllowed(usdAttr)
-        self.assertFalse(isAttrEditAllowed, False)
+        self.assertFalse(isAttrEditAllowed)
```
Correct, the logic in the "if isAttrEditAllowed:" conditions are dead code. 
When these test were written, we didn't have such restriction to block attribute edits but now we have. In the case of top_layer.ma example, the top layer "Ball.maya.usd" has already set certain attributes ( e.g doubleSided, etc... ) and setting them in the weaker layers are blocked now and therefore the following tests after "if isAttrEditAllowed:" condition results in error.

ppt-adsk replied:
Right --- and my suggestion here, to keep the test code coverage for the UFE attributes support, is to set the attributes in the stronger layer, otherwise we won't be testing the Ufe::Attribute::set() interfaces any more.  You've already added two new tests to test attribute blocking, so we don't need additional testing for that, but we do need to keep the test coverage for Ufe::Attribute::set().

HamedSabri-adsk replied:
Fair enough, let me have a look.

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -80,6 +80,24 @@ bool stringBeginsWithDigit(const std::string& inputString)
+
+    auto iter = std::find(std::begin(layerStack), std::end(layerStack), layer);
+
+    if (iter != layerStack.end()) {
```
addressed ac2bb4dc732e9862957b25a17eb9f6f7719adcf3

ppt-adsk replied:
Think we're settling on TF_ debugging facilities, e.g. TF_AXIOM(iter != layerStack.end())

HamedSabri-adsk replied:
Yup that's what I tried first and spent 5 mins last night scratching my head why boost python doesn't like it:

```
15:   KERNELBASE.dll!DebugBreak
15: 00007FFA24C99E10Result: C:/Users/sabrih/Desktop/workspace/build/RelWithDebInfo/test/testSamples/ballset/StandaloneScene/top_layer.ma
15: Fatal Error. Attempting to save in C:/Users/sabrih/AppData/Local/Temp/sabrih.20210318.1311.ma
```
I tried to search USD code base and I can't find any examples where TF_AXIOM is being used in any functions that are exposed to Boost python. 

I believe the good old assert() should do fine here.

ppt-adsk replied:
Again, Python shouldn't have anything to do this this...  Is the TF_AXIOM failing then?

HamedSabri-adsk commented on test/lib/ufe/testAttribute.py:
```diff
@@ -417,19 +395,15 @@ def testAttributeFloat3(self):
+        self.assertVectorAlmostEqual(ufeAttr.get(), usdAttr.Get())
+
+        self.runUndoRedo(ufeAttr,
+                         ufe.Vector3f(vec.x()+1.0, vec.y()+2.0, vec.z()+3.0))
```
There is no more dead code and all my silly isAttrEditAllowed() checks are removed.

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -410,29 +423,36 @@ bool isAttributeEditAllowed(const PXR_NS::UsdAttribute& attr, std::string* errMs
-        auto topLayer = attr.GetPropertyStack().front()->GetLayer();
+        // compare the calculated index between the "attr" and "edit target" layers.
+        if (findLayerIndex(prim, strongestLayer) < targetLayerIndex) {
+            if (errMsg) {
```
Thank you Pierre. Indeed we needed to check for the null here....

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -410,29 +423,36 @@ bool isAttributeEditAllowed(const PXR_NS::UsdAttribute& attr, std::string* errMs
 
     if (!propertyStack.empty()) {
+        // get the strongest layer that has the attr.
+        auto strongestLayer = attr.GetPropertyStack().front()->GetLayer();
```
renamed topLayer with strongestLayer

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -80,19 +82,30 @@ bool stringBeginsWithDigit(const std::string& inputString)
+                return position;
+            }
+            ++position;
+        }
```
The original algorithm that I wrote here was very naive and didn't handle the cases where the attribute came in across composition arcs. By walking the expanded primIndex, we would be able to visit all the "**sites**" and their "**local layer stack**".

I am also thinking of another way to write this code but believe for the pass should be good. 

Thanks to Brett Levin and Spiff to help me better understand the "primindex" and "site" concepts.

ppt-adsk replied:
So this linearizes the complete layer stack and expands out each site, such that if there were m node sites each with n layers, then the value of position at the end of the loop would be (m x n) - 1, right?

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -25,6 +25,8 @@
 #include <pxr/base/tf/hashset.h>
 #include <pxr/base/tf/stringUtils.h>
+#include <pxr/usd/pcp/layerStack.h>
+#include <pxr/usd/pcp/site.h>
```
bring in "pcp/site" and "pcp/layerStack".

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -410,29 +423,36 @@ bool isAttributeEditAllowed(const PXR_NS::UsdAttribute& attr, std::string* errMs
+    // TODO: "Value Clips" are UsdStage-level feature, unknown to Pcp.So if the attribute in
+    // question is affected by Value Clips, we would will likely get the wrong answer. See Spiff
+    // comment for more information :
     // https://groups.google.com/g/usd-interest/c/xTxFYQA_bRs/m/lX_WqNLoBAAJ
```
See my comment in regard to "Value Clips".

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -80,19 +82,30 @@ bool stringBeginsWithDigit(const std::string& inputString)
+    const PXR_NS::PcpPrimIndex& primIndex = prim.GetPrimIndex();
+
+    // iterate through the expanded primIndex
+    for (const PcpNodeRef& node : primIndex.GetNodeRange()) {
```
PF results back and I am hit with `-Wrange-loop-analysis` with Clang on MacOS:

```
    /local/S/jenkins/workspace/ecg-mayausd-branch-preflight-2022-python3-osx/ecg-maya-usd/maya-usd/lib/mayaUsd/ufe/Utils.cpp:94:28: error: loop variable 'node' is always a copy because the range of type 'pxrInternal_v0_20__pxrReserved__::PcpNodeRange' (aka 'pair<pxrInternal_v0_20__pxrReserved__::PcpNodeIterator, pxrInternal_v0_20__pxrReserved__::PcpNodeIterator>') does not return a reference [-Werror,-Wrange-loop-analysis]
        for (const PcpNodeRef& node : primIndex.GetNodeRange()) {
```

I believe I have previously commented on this flag before in https://github.com/Autodesk/maya-usd/pull/930/ 

My original intention here is that I don't want a copy in my range loop and find it hard this iterator object is always copy. 

`for(const PcpNodeRef &node: primIndex.GetNodeRange()) `

Not to mention that this is pattern that is used everywhere in USD code base:

e.g https://github.com/PixarAnimationStudios/USD/blob/dc710925675f7f58f7a37f6c18d046a687b09271/pxr/usd/usdShade/material.cpp#L439

ppt-adsk replied:
Did you try const auto& node?

HamedSabri-adsk commented on lib/mayaUsd/ufe/Utils.cpp:
```diff
@@ -80,19 +82,30 @@ bool stringBeginsWithDigit(const std::string& inputString)
-// find position index for a given layer in the local LayerStack ( strong-to-weak order )
-uint32_t findLayerIndex(const UsdStageWeakPtr& stage, const PXR_NS::SdfLayerHandle& layer)
+// This function calculates the position index for a given layer across all
+// the site's local LayerStack.
```
@ppt-adsk  Yes it is. Every Site ( simply a PcpNode in the DAG ) has a "local layer stack" and "scene path".

"Local" just means a site's "root layer and all its sublayers". That's the sense in which is a "local" layer stack: it has no "global" knowledge.

To make it easier to see things, I make a change in Pcp DAG graph representation, where I show the size of
each layer stack in each site. 

![PrimIndex_Concept](https://user-images.githubusercontent.com/48299423/112158652-c16b1680-8bbe-11eb-83eb-982a5800125f.jpg)

ppt-adsk replied:
Fantastic explanation, thanks!  Therefore, shouldn't the comment read "across all the site's local LayerStacks" plural layer stacks?

HamedSabri-adsk replied:
Good point, I see. Yes should be plural.

HamedSabri-adsk replied:
Addressed in 818e642effab378839cf042abf4af9b6462df5a6

HamedSabri-adsk commented:
> So this linearizes the complete layer stack and expands out each site, such that if there were m node sites each with n layers, then the value of position at the end of the loop would be (m x n) - 1, right?

Yes and no. The number of iteration is (M x N)

where 

M = total number of sites
N = total number of layers 
Position = N - 1

It is because N -1 because position value is incremented after the check.

So let's use the DAG graph above as an example. By the way, there is a tiny typo is my above image. **Site 5** should be replaced by **Site 6**. Apologizes.

M = 6 
N = 10

If I am searching for **@/Ball.maya.usd@</Ball/mesh>**, the position would be **9**

Site 1
-------
top_layer-session.usda
top_layer.usda
Assembly_room_set.usda
edits.usda
edits.usda

Site 2
-------
Ball.usd

Site 3
-------
Ball_payload.usd

Site 4
-------
Ball.shadingVariants.usda

Site 5
-------
Ball.shadingVariants.usda

Site 6
-------
Ball.maya.usd

HamedSabri-adsk commented:
> Did you try const auto& node?

@ppt-adsk I haven't tried but one shouldn't have to...  

As I reported before `range-loop-analysis` has a tendency to give false positive results. We don't have this flag in Maya and I also haven't seen it enforced in other projects that I know.

**Avoid Wrange-loop-analysis false positives**

https://reviews.llvm.org/rG318677e78def0023d210a29f4b3cf648e02f9fdc

https://quuxplusone.github.io/blog/2020/08/26/wrange-loop-analysis/

ppt-adsk reviewed (commented):
Will leave it up to you as to how to fix the compiler warning as error on macOS, no particular guidance I could give on that one.

HamedSabri-adsk commented:
@ppt-adsk per our internal discussion. 

I reverted the clang flag change to minimize the impact on others who may need it:

b89e2b295bc3277d7af2238f799e21064c35edad

Dropped `const&` for PcpNodeRef and use it by value since it is just a pointer.  We could also use the `auto&& ` pattern but going by value should be more obvious for some. Added my Godbolt example here for reference:

https://gcc.godbolt.org/z/cK3bWPT4d
  

b2356d4f806b6ffa02c35278059cb252b7e84656
